livre/eq/Channel.cpp

```diff
@@ -515,68 +468,155 @@ class Channel
             frame->disableBuffer( eq::Frame::BUFFER_DEPTH );
-            frame->getFrameData()->setRange( _drawRange );
-        }
-    }
-
-    bool composeOnly() const
-    {
-        return _drawRange == eq::Range::ALL;
     }
 
     void frameAssemble( const eq::Frames& frames )
     {
         eq::PixelViewport coveredPVP;
         eq::Frames dbFrames;
-        eq::Zoom zoom( eq::Zoom::NONE );
 
         // Make sure all frames are ready and gather some information on them
-        prepareFramesAndSetPvp( frames, dbFrames, coveredPVP, zoom );
-
-        if( dbFrames.empty( ))
-            return;
-
+        prepareFramesAndSetPvp( frames, dbFrames, coveredPVP );
         coveredPVP.intersect( _channel->getPixelViewport( ));
 
-        // calculate correct frames sequence
-        eq::FrameDataPtr data = _frame.getFrameData();
+        if( dbFrames.empty() || !coveredPVP.hasArea( ))
+            return;
 
-        if( !composeOnly() && coveredPVP.hasArea( ))
+        if( useDBSelfAssemble( )) // add self to determine ordering
         {
+            eq::FrameDataPtr data = _frame.getFrameData();
             _frame.clear( );
+            _frame.setOffset( eq::Vector2i( 0, 0 ));
             data->setRange( _drawRange );
+            data->setPixelViewport( coveredPVP );
             dbFrames.push_back( &_frame );
         }
 
-        // Update range
-        eq::Range newRange( 1.f, 0.f );
+        orderFrames( dbFrames, computeModelView( ));
+
+        if( useDBSelfAssemble( )) // read back self frame
+        {
+            if( dbFrames.front() == &_frame ) // OPT: first in framebuffer!
+                dbFrames.erase( dbFrames.begin( ));
+            else
+            {
+                _frame.readback( _channel->getObjectManager(),
+                                 _channel->getDrawableConfig(),
+                                 _channel->getRegions( ));
+                clearViewport( coveredPVP );
+                // offset for assembly
+                _frame.setOffset( eq::Vector2i( coveredPVP.x, coveredPVP.y ));
+            }
+        }
+
+        LBINFO << "Frame order: ";
+        for( const eq::Frame* frame : dbFrames )
+            LBINFO << frame->getName() <<  " "
+                   << frame->getFrameData()->getRange() << " : ";
+        LBINFO << std::endl;
+
+        try // blend DB frames in computed order
+        {
+            eq::Compositor::assembleFramesSorted( dbFrames, _channel, 0,
+                                                  true /* blendAlpha */ );
+        }
+        catch( const std::exception& e )
+        {
+            LBWARN << e.what() << std::endl;
+        }
+
+        // Update draw range
         for( size_t i = 0; i < dbFrames.size(); ++i )
+            _drawRange.merge( dbFrames[i]->getRange( ));
+    }
+
+    bool useDBSelfAssemble() const { return _drawRange != eq::Range::ALL; }
+
+    static bool cmpRangesInc(const eq::Frame* a, const eq::Frame* b )
```

Comment:
It's a separate, non-functional, commit for sort-last compositing. Imo we can merge this as well since it improves the current situation. I did not mention it in the PR since it's not working yet.

What's you opinion?